crates/test-programs/build.rs

```diff
@@ -178,6 +235,22 @@ mod wasi_tests {
                         "truncation_rights" => true,
                         "path_link" => true,
                         "dangling_fd" => true,
+                        // TODO: virtfs files cannot be poll_oneoff'd yet
+                        "poll_oneoff_virtualfs" => true,
+                        // TODO: virtfs does not support filetimes yet.
+                        "path_filestat_virtualfs" |
+                        "fd_filestat_set_virtualfs" => true,
+                        // TODO: virtfs does not support symlinks yet.
+                        "nofollow_errors_virtualfs" |
+                        "path_link_virtualfs" |
+                        "readlink_virtualfs" |
+                        "readlink_no_buffer_virtualfs" |
+                        "dangling_symlink_virtualfs" |
+                        "symlink_loop_virtualfs" |
+                        "path_symlink_trailing_slashes_virtualfs" => true,
+                        // TODO: virtfs does not support rename yet.
+                        "path_rename_trailing_slashes_virtualfs" |
+                        "path_rename_virtualfs" => true,
```

Comment:
Hmm, this looks like we're redoing every `VirtualFs` test on every host. Am I correct in assuming so, and do we actually need to do that given that our de facto host is "virtual"? How much effort do you think it would take to have `VirtualFs` tests behind a feature flag (treating it as a separate "host" altogether), and do you reckon it would actually make sense?

Reply:
So.. my thinking here is that because `Descriptor` involves non-virtual variants, in the strictest sense we _ought_ to run on all hosts. Slightly less interesting, we'd need to pick _some_ host to run `*_virtualfs` tests on, unless we codify the idea of a "virtual" host to the point of having it define "OsHandle" for tests too.

Realistically, your recommendation to adjust dispatch to `VirtualFs` to consistently be before getting reaching host-specific impls would help here. At that point I think running these on only one host (linux, because I'm biased?) is less risky .. when doing full CI runs, anyway.

I think the last reason to run `VirtualFs` tests for each host is local development; Running on every host is helpful for local platform-specific development, since you'll only run one host's view. Do you think it makes sense to make it conditional on `tests are running in CI`?

Reply:
Yeah, I see what you mean by having a mix of the two: a host plus some virtual fs on top of it. The reason I mentioned it here is since there seems to be a lot of code duplication so I was trying to figure out if we could somehow reduce it especially since it seems to be very much copy-paste reuse. 🤔 

> I think the last reason to run VirtualFs tests for each host is local development; Running on every host is helpful for local platform-specific development, since you'll only run one host's view. Do you think it makes sense to make it conditional on tests are running in CI?

I think I'd actually be more comfortable always having it on so that we make sure we don't break it with any changes to `wasi-common` in the future. In fact, I think that full test coverage that's always on trumps some code duplication here and there. But it would be nice to have the first without the second ;-)